internal/indexer/linux/coalescer.go

```diff
@@ -71,101 +69,142 @@ func (c *Coalescer) Coalesce(ctx context.Context, layers []*claircore.Layer) (*c
+					}
+					dbs[pkg.PackageDB].packages[pkg.ID] = pkg
+					dbs[pkg.PackageDB].environments[pkg.ID] = environment
+				}
 			}
 		}
 	}
-
-}
-
-// prune removes packages which do not exist in the "newest" verion of a package database.
-//
-// this assumes we are working with linux distribution package managers and an addition or
-// removal of a package presents the entire database in a layer due to CoW semantics.
-func (c *Coalescer) prune(ctx context.Context, artifacts []layerArtifacts) {
 	if ctx.Err() != nil {
-		return
+		return ctx.Err()
 	}
-
-	keep := map[string]map[int]struct{}{}
-
-	// walk layer artifacts backwards searching for newest package databases
+	// we now have all the packages associated with their introduced in layers and environments.
+	// we must now prune any packages removed between layers. this coalescer works on the assumption
+	// that any changes to a package's database (dpkg, rpm, alpine, etc...) causes the entire database
+	// file to be written to the layer in which the change occurs. this assumption therefore
+	// allows for the following algorithm
+	// 1) walk layers backwards searching for newest modification of package dabatase.
+	// 2) if we encounter a package existing in a particular database it means all packages within this package database are present.
+	//    record all packages found into a temporary map.
+	//    when we are finished searching the current layer add a key/value to the penultimate map indicating
+	//    we no longer care about this set of package databases.
+	// 3) continue for all layers, always checking to see if we've already encountered a package database.
+	//    as we only want to inventory packages from the newest package database
+	// 4) once all layers are scanned begin removing package ids not present in our penultimate packagesToKeep map
+	var packagesToKeep = map[string][]int{}
 	for i := len(artifacts) - 1; i >= 0; i-- {
-		a := artifacts[i]
-		if len(a.pkgs) == 0 {
+		layerArtifacts := artifacts[i]
+		if len(layerArtifacts.pkgs) == 0 {
 			continue
 		}
-
-		// split package array into a per-package-db representation
-		packageDBs := map[string][]*claircore.Package{}
-		for _, pkg := range a.pkgs {
-			packageDBs[pkg.PackageDB] = append(packageDBs[pkg.PackageDB], pkg)
-		}
-
-		// for each package db discovered check if we've seen it. if we haven't
-		// seen it record all packages in this package db
-		for db, packages := range packageDBs {
-			if _, ok := keep[db]; !ok {
-				keep[db] = map[int]struct{}{}
-				for _, pkg := range packages {
-					keep[db][pkg.ID] = struct{}{}
+		// used as a temporary accumulator of package ids in this layer
+		var tmpPackagesToKeep = map[string][]int{}
+		for _, pkg := range layerArtifacts.pkgs {
+			// have we already inventoried packages from this database ?
+			if _, ok := packagesToKeep[pkg.PackageDB]; !ok {
+				// ... we haven't so add to our temporary accumlator
+				if _, ok := tmpPackagesToKeep[pkg.PackageDB]; !ok {
```

Comment:
I think this would be clearer without the if/else:

```
tk := tmpPackagesToKeep[pkg.PackageDB]
tmpPackagesToKeep[pkg.PackageDB] = append(tk, pkg.ID)
```

It's legal to append to a nil slice, so the first call would just create one.